app/models/collection.rb

```diff
@@ -162,6 +162,54 @@ def has_untranscribed_pages?
     self.works.joins(:work_statistic).where('work_statistics.total_pages > work_statistics.transcribed_pages').exists?
   end
 
+  def update_works_stats
+    works = self.works.includes(:work_statistic)
+    works_stats = get_works_stats_hash
+    works.each do |w|
+      w.work_statistic.recalculate_from_hash(works_stats[w.id]) if works_stats.key?(w.id)
+    end
+    calculate_complete
+  end
+
+  def enable_ocr
+    works.update_all(ocr_correction: true)
+    update_works_stats
+  end
+
+  def disable_ocr
+    works.update_all(ocr_correction: false)
+    update_works_stats
+  end
+
+  def get_works_stats_hash
+    stats = {}
+    work_prototype = {
+      transcription: {},
+      translation: {},
+      total: 0
+    }
+    work_ids = works.ids
+    transcription = Page.where(work_id: work_ids).group(:work_id, :status).count
+    translation = Page.where(work_id: work_ids).group(:work_id, :translation_status).count
+    totals = Page.where(work_id: work_ids).group(:work_id).count
+    
+    transcription.each do |(id, status), value|
+      stats[id] = work_prototype unless stats.has_key?(id)
+      stats[id][:transcription][status] = value
+    end
+
+    translation.each do |(id, status), value|
+      stats[id] = work_prototype unless stats.has_key?(id)
+      stats[id][:translation][status] = value
+    end
+
+    totals.each do |id, value|
+      stats[id] = work_prototype unless stats.has_key?(id)
+      stats[id][:total] = value
+    end
+    stats
+  end
```

Comment:
I'd be really tempted to make this block more concise like this:
```
work_ids.each do |work_id|
  stats[id]=work_prototype  # consider a copy/clone/dup if this is immutable or we are overwriting it?
  stats[id][:total]=transcription[id]
  stats[id][:translation]=translation[id]
  stats[id][:total]=totals[id]
end
```

It's just style, but I think it makes it a bit clearer.

Reply:
I think there's actually a bit more going on here that you may have missed. The resulting values of the `transcription` and `translation` queries are grouped by `work.id` AND by `work.status` to the result is a somewhat de-normalized list of counts like this with the format `[work_id, status] => count` :
```
[15, 'needs_review'] => 5
[15, 'transcribed'] => 10
... 
```
So, it's not a nested hash, like you'd expect (I think this is what your example is assuming):
```
15 => {
  'needs_review' => 5,
  'transcribed' => 10
}
```
I certainly could rewrite the code as the above, but I'd need to write a helper function to convert the query results into the right format first.